Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp

```diff
@@ -2366,42 +2365,52 @@ RefPtr<ByteArrayPixelBuffer> CanvasRenderingContext2DBase::cacheImageDataIfPossi
         .rows = cachedBuffer->data().data(),
     };
     convertImagePixels(source, destination, size);
-
-    m_cachedImageData.emplace(*this, *cachedBuffer);
+    m_cachedContents.emplace<CachedContentsImageData>(*this, *cachedBuffer);
     return cachedBuffer;
 }
 
-RefPtr<ImageData> CanvasRenderingContext2DBase::takeCachedImageDataIfPossible(const IntRect& sourceRect, PredefinedColorSpace colorSpace) const
+RefPtr<ImageData> CanvasRenderingContext2DBase::makeImageDataIfContentsCached(const IntRect& sourceRect, PredefinedColorSpace colorSpace) const
 {
-    if (!m_cachedImageData)
-        return nullptr;
-
-    if (sourceRect != IntRect { { }, canvasBase().size() })
-        return nullptr;
-
-    if (canvasBase().size() != m_cachedImageData->imageData->size())
-        return nullptr;
-
-    if (colorSpace != m_settings.colorSpace)
-        return nullptr;
-
-    Ref pixelBuffer = WTFMove(m_cachedImageData->imageData);
-    m_cachedImageData = std::nullopt;
-    auto size = pixelBuffer->size();
-    auto data = pixelBuffer->takeData();
-    unsigned bytesPerRow = static_cast<unsigned>(size.width()) * 4u;
-    ConstPixelBufferConversionView source {
-        .format = pixelBuffer->format(),
-        .bytesPerRow = bytesPerRow,
-        .rows = data->data(),
-    };
-    PixelBufferConversionView destination {
-        .format = { AlphaPremultiplication::Unpremultiplied, PixelFormat::RGBA8, pixelBuffer->format().colorSpace },
-        .bytesPerRow = bytesPerRow,
-        .rows = data->data(),
-    };
-    convertImagePixels(source, destination, size);
-    return ImageData::create(size, WTFMove(data), m_settings.colorSpace);
+    return WTF::switchOn(m_cachedContents,
+        [&](CachedContentsTransparent&)  -> RefPtr<ImageData> {
+            auto imageData = ImageData::create(sourceRect.size(), colorSpace);
+            if (imageData)
+                imageData->data().zeroFill();
+            return imageData;
+        },
+        [](CachedContentsUnknown&) -> RefPtr<ImageData> {
+            return nullptr;
+        },
+        [&](CachedContentsImageData& cachedImageData) -> RefPtr<ImageData> {
+            // Always consume the cache.
+            Ref pixelBuffer = WTFMove(cachedImageData.imageData);
+            m_cachedContents.emplace<CachedContentsUnknown>();
```

Comment:
I always find this kind of pattern a bit worrying, since it leaves us with a dangling `CachedContentsImageData& cachedImageData`. I don't have a good suggestion on what to do instead, unless we change to use something other than `WTF::switchOn`.

Reply:
What about the ifs in the new patch